docs/aiken-integration.md

```diff
@@ -0,0 +1,516 @@
+                                }
+                                1 -> when fields is {
+                                    [] -> MyNothing
+                                    _ -> error @"Bad parse"
+                                }
+                                _ -> error @"Bad parse"
+                            }
+                    })(theData), (fn(_theData) { error @"Bad parse" })(theData), (fn(_theData) { error @"Bad parse"})(theData), (fn(_theData){error @"Bad parse"})(theData), (fn(_theData) { error @"Bad parse"})(theData))
+        answer
+    }
+}
+
+pub fn toData(toDataT : fn(t) -> Data) -> fn(MyOption<t>) -> Data {
+    fn(theOption) {
+        when theOption is {
+            MyJust(justValue) -> builtin.constr_data(0, [toDataT(justValue)])
+            MyNothing -> builtin.constr_data(1, [])
+        }
+    }
+}
+
+// Example usages:
+
+test from_nothing_test() {
+    fromData(lb_prelude_int.fromData)(builtin.constr_data(1, [])) == MyNothing
+}
+
+test from_just_test() {
+    fromData(lb_prelude_int.fromData)(builtin.constr_data(0, [builtin.i_data(69)])) == MyJust(69)
+}
+
+test from_to_nothing_test() {
+    toData(lb_prelude_int.toData)(fromData(lb_prelude_int.fromData)(builtin.constr_data(1, []))) == builtin.constr_data(1, [])
+}
+
+test from_to_just_test() {
+    toData(lb_prelude_int.toData)(fromData(lb_prelude_int.fromData)(builtin.constr_data(0, [builtin.i_data(69)]))) == builtin.constr_data(0, [builtin.i_data(69)])
+}
+```
+
+This translation of type classes has some limitations such as:
+
+* All type class instances must be defined in the same module that the type is defined in i.e., orphan instances are forbidden.
+
+* Only Haskell2010 type classes would be supported.
+
+### Aiken's encoding of its data is different from LambdaBuffers encoding
+
+All onchain scripts must be compiled to UPLC which must in some method represent the higher level language constructs like data types in the original language.
+Often, data types in a higher level language are translated to UPLC's builtin `Data` type which supports types like lists, constructors, integers, bytestrings, and maps.
+Note that data which will exist as a datum or redeemer must admit a representation with this `Data` type.
+
+LambdaBuffers chooses a particularly efficient encoding of its data types to `Data` mapping to its target languages that map to UPLC.
+For example, a record like
+
+```purescript
+record MyRecord = { a : Integer, b : Integer }
+```
+
+would be translated to
+
+```purescript
+[a, b]
+```
+
+i.e., records are lists of all record components[^recordsSpecialCases].
+
+[^recordsSpecialCases]: Singleton records are encoded as just the single element.
+
+If LambdaBuffers compiled `MyRecord` to a [record in Aiken](https://aiken-lang.org/language-tour/custom-types) as follows
+
+```rust
+type MyRecord {
+    a : Int,
+    b : Int
+}
+```
+
+we know that Aiken will internally represent this as the following `Data` type
```

Comment:
Emphasizing on "we know" being "we've observed" because docs don't mention them and push you to use the json schema thingie